library/std/src/thread/mod.rs

```diff
@@ -1264,23 +1276,54 @@ impl ThreadId {
 // Thread
 ////////////////////////////////////////////////////////////////////////////////
 
-/// The internal representation of a `Thread`'s name.
-enum ThreadName {
-    Main,
-    Other(CString),
-    Unnamed,
-}
+/// The parker for the main thread. This avoids having to allocate an Arc in `fn main() {}`.
+static MAIN_PARKER: SyncUnsafeCell<MaybeUninit<Parker>> =
+    SyncUnsafeCell::new(MaybeUninit::uninit());
 
-/// The internal representation of a `Thread` handle
-struct Inner {
-    name: ThreadName, // Guaranteed to be UTF-8
+/// The internal representation of a `Thread` that is not the main thread.
+struct OtherInner {
+    name: Option<CString>, // Guaranteed to be UTF-8
     id: ThreadId,
     parker: Parker,
 }
 
+/// The internal representation of a `Thread` handle.
+#[derive(Clone)]
+enum Inner {
+    Main,
+    Other(Pin<Arc<OtherInner>>),
+}
+
 impl Inner {
-    fn parker(self: Pin<&Self>) -> Pin<&Parker> {
-        unsafe { Pin::map_unchecked(self, |inner| &inner.parker) }
+    fn id(&self) -> ThreadId {
+        match self {
+            Self::Main => ThreadId::new_main(),
+            Self::Other(other) => other.id,
+        }
+    }
+
+    fn name(&self) -> Option<&CStr> {
+        match self {
+            Self::Main => Some(c"main"),
+            Self::Other(other) => other.name.as_deref(),
+        }
+    }
+
+    fn parker(&self) -> Pin<&Parker> {
+        match self {
+            Self::Main => {
+                // Safety: MAIN_PARKER only ever has a mutable reference when Inner::Main is initialised.
+                let static_ref: &'static MaybeUninit<Parker> = unsafe { &*MAIN_PARKER.get() };
```

Comment:
I would like some more wording in this safety comment. I *think* this is ok, but it requires some more advanced thinking. Mainly because of this:
https://www.rustexplorer.com/b#LyoKW2RlcGVuZGVuY2llc10KY3RvciA9ICIwLjIuOCIKKi8KCiNbY3Rvcjo6Y3Rvcl0KZm4gaW5pdCgpIHsKICAgIHN0ZDo6dGhyZWFkOjpzcGF3bih8fCBwcmludGxuISgiSGVsbG8gV29ybGQhIikpOwp9CgpmbiBtYWluKCkgewogICAgcHJpbnRsbiEoIm9vcHMgaXMgYW0gc28gaW5pdCIpOwp9

While it's true that it only has a mutable reference when main is initialized, the initialization can run in parallel with a different thread (is my program with a life-before-main thread guaranteed to work? i don't think so. is it sound? who knows. but it sure would be nice if it didn't cause unsoundness here :3).
So, I thought whether it could happen that someone reads this concurrently with the main thread initializing, which would obviously be bad.

I think the answer to that is no. Because the only way to read from MAIN_PARKER is to via this function, and this function requires a `&Inner` of the main thread, and the only way to get that is with the main thread giving such a reference out itself, which means it has been initialized.

Could you reflect this reasoning a bit better in the comment? I suggest

```suggestion
                // Safety: MAIN_PARKER is only ever read in this function, which requires access
                // to an existing `&Inner` value, which can only be accessed via the main thread
                // giving away such an instance from `current()`, implying that initialization,
                // the only write to `MAIN_PARKER`, has been completed.
                let static_ref: &'static MaybeUninit<Parker> = unsafe { &*MAIN_PARKER.get() };
```